File: algorithms/0728.self-dividing-numbers/solution.rb
# @param {Integer} left
# @param {Integer} right
# @return {Integer[]}
def self_dividing_numbers(left, right)
  res = []

  (left..right).each do |n|
    next res << n if n < 10

    divisor = n.to_s.chars.map(&:to_i)
    self_dividing = divisor.each do |d|
      if d != 0 && n % d == 0
        true
      else
        break false
      end
    end
    res << n if self_dividing
  end

  res
end

File: algorithms/0728.self-dividing-numbers/solution.rs
impl Solution {
    pub fn self_dividing_numbers(left: i32, right: i32) -> Vec<i32> {
        let mut res: Vec<i32> = vec![];

        for n in left..=right {
            if n < 10 {
                res.push(n);
                continue;
            }

            let divisor = n
                .to_string()
                .chars()
                .map(|c| c as i32 - 48)
                .collect::<Vec<i32>>();

            let mut self_dividing = false;
            for d in divisor {
                if d != 0 && n % d == 0 {
                    self_dividing = true;
                } else {
                    self_dividing = false;
                    break;
                }
            }

            if self_dividing {
                res.push(n);
            }
        }

        res
    }
}
